src/funnels/Depositor.sol

```diff
@@ -0,0 +1,275 @@
+    {
+        require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
+
+        PairLimit memory limit = limits[p.gem0][p.gem1];
+        require(block.timestamp >= limit.zzz + limit.hop, "Depositor/too-soon");
+        limits[p.gem0][p.gem1].zzz = uint64(block.timestamp);
+
+        UniV3PoolLike pool = _getPool(p.gem0, p.gem1, p.fee);
+        liquidity = (p.liquidity == 0)
+            ? _getLiquidityForAmts(pool, p.tickLower, p.tickUpper, p.amt0Desired, p.amt1Desired)
+            : p.liquidity;
+
+        (amt0, amt1) = pool.mint({
+            recipient: address(this),
+            tickLower: p.tickLower,
+            tickUpper: p.tickUpper,
+            amount   : liquidity,
+            data     : abi.encode(MintCallbackData({gem0: p.gem0, gem1: p.gem1, fee: p.fee}))
+        });
+        require(amt0 >= p.amt0Min && amt1 >= p.amt1Min, "Depositor/exceeds-slippage");
+        require(amt0 <= limit.cap0 && amt1 <= limit.cap1, "Depositor/exceeds-cap");
+
+        emit Deposit(msg.sender, p.gem0, p.gem1, liquidity, amt0, amt1);
+    }
+
+    function withdraw(LiquidityParams memory p, bool takeFees)
+        external
+        auth
+        returns (uint128 liquidity, uint256 amt0, uint256 amt1)
+    {
+        require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
+
+        PairLimit memory limit = limits[p.gem0][p.gem1];
+        require(block.timestamp >= limit.zzz + limit.hop, "Depositor/too-soon");
+        limits[p.gem0][p.gem1].zzz = uint64(block.timestamp);
+
+        UniV3PoolLike pool = _getPool(p.gem0, p.gem1, p.fee);
+        liquidity = (p.liquidity == 0)
+            ? _getLiquidityForAmts(pool, p.tickLower, p.tickUpper, p.amt0Desired, p.amt1Desired)
+            : p.liquidity;
+
+        (amt0, amt1) = pool.burn({ tickLower: p.tickLower, tickUpper: p.tickUpper, amount: liquidity });
+        require(amt0 >= p.amt0Min  && amt1 >= p.amt1Min,  "Depositor/exceeds-slippage");
+        require(amt0 <= limit.cap0 && amt1 <= limit.cap1, "Depositor/exceeds-cap");
+
+        (uint256 collected0, uint256 collected1) = pool.collect({
+            recipient       : buffer,
+            tickLower       : p.tickLower,
+            tickUpper       : p.tickUpper,
+            amount0Requested: takeFees ? type(uint128).max : uint128(amt0),
+            amount1Requested: takeFees ? type(uint128).max : uint128(amt1)
+        });
+
+        emit Withdraw(msg.sender, p.gem0, p.gem1, p.liquidity, amt0, amt1, collected0, collected1);
+    }
+
+    struct CollectParams {
+        address gem0;
+        address gem1;
+        uint24  fee;
+        int24   tickLower;
+        int24   tickUpper;
+    }
+
+    function collect(CollectParams memory p)
+        external
+        auth
+        returns (uint256 collected0, uint256 collected1)
+    {
+        require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
+
+        UniV3PoolLike pool = _getPool(p.gem0, p.gem1, p.fee);
+        pool.burn(p.tickLower, p.tickUpper, 0); // update the position's owed fees
```

Comment:
I'd use the same format than in `withdraw` for passing the params.

Reply:
Done.